agent/proxycfg/manager.go

```diff
@@ -40,49 +40,67 @@ type Manager struct {
 	stateCh chan struct{}
 
 	mu       sync.Mutex
-	proxies  map[string]*State
+	proxies  map[string]*state
 	watchers map[string]map[uint64]chan *ConfigSnapshot
 }
 
+// ManagerConfig holds the required external dependencies for a Manager
+// instance. All fields must be set to something valid or the manager will
+// panic.
+type ManagerConfig struct {
+	// Cache is the agent's cache instance that can be used to retreive, store and
+	// monitor state for the proxies.
+	Cache *cache.Cache
+	// state is the agent's local state to be watched for new proxy registrations.
+	State *local.State
+	// source describes the current agent's identity, it's used directly for
+	// prepared query discovery but also indirectly as a way to pass current
+	// Datacenter name into other request types that need it. This is sufficient
+	// for now and cleaner than passing the entire RuntimeConfig.
+	Source *structs.QuerySource
+	// logger is the agent's logger to be used for logging logs.
+	Logger *log.Logger
+}
+
 // NewManager constructs a manager from the provided agent cache.
-func NewManager(logger *log.Logger, cache *cache.Cache, state *local.State,
-	source *structs.QuerySource) *Manager {
+func NewManager(cfg ManagerConfig) (*Manager, error) {
+	if cfg.Cache == nil || cfg.State == nil || cfg.Source == nil ||
+		cfg.Logger == nil {
+		return nil, errors.New("all ManagerConfig fields must be provided")
+	}
 	m := &Manager{
-		logger: logger,
-		cache:  cache,
-		state:  state,
-		source: source,
+		cfg: cfg,
 		// Single item buffer is enough since there is no data transferred so this
 		// is "level triggering" and we can't miss actual data.
 		stateCh:  make(chan struct{}, 1),
-		proxies:  make(map[string]*State),
+		proxies:  make(map[string]*state),
 		watchers: make(map[string]map[uint64]chan *ConfigSnapshot),
 	}
-	go m.run()
-	return m
+	return m, nil
 }
 
-// run is the internal state syncing loop that watches the local state for
-// changes.
-func (m *Manager) run() {
-
+// Run is the long-running method that handles state synching. It should be run
+// in it's own goroutine and will continue until a fatal error is hit or Close
+// is called.
+func (m *Manager) Run() error {
```

Comment:
Can this be run multiple times safely? What happens if Run is called again while already running (before Close)? Those are things I like to have clarified in these docs.

Reply:
Reasonable. I handled the case Run is called after close but yeah will do bad things if multiple concurrent Runs are called. I'll document that.